### PR TITLE
add C code for graph

### PR DESCRIPTION
**Pull request by NI-SW**
If this PR is related to coding or code translation, please fill out the checklist and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

**Comment by NI-SW** [comment]
XD

File: codes/c/chapter_graph/graph_adjacency_matrix.c [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+ * Created Time: 2023-07-06
+ * Author: NI-SW (947743645@qq.com)
+ */
+
+#include "../utils/common.h"
+
+/* 基于邻接矩阵实现的无向图类结构 */
+typedef struct graphAdjMat graphAdjMat;
+struct graphAdjMat
+{
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_matrix.c:12]: 注意代码风格统一，格式化详见[14.1.3.   C/C++ 环境](https://www.hello-algo.com/chapter_appendix/installation/#1413-cc)

File: codes/c/chapter_graph/graph_adjacency_matrix.c [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+                t->adjMat[i][j] = t->adjMat[i][j + 1];            
+            }
+        }
+    }
+    t->sizei--;
+}
+
+/* 打印顶点与邻接矩阵 */
+void print_graph(graphAdjMat* t)
+{
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_matrix.c:164]: 命名不一致

File: codes/c/chapter_graph/graph_adjacency_matrix.c [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    }
+    printf("\n邻接矩阵:\n");
+    for(int i = 0;i < t->size(t->self); i++)
+    {
+        for(int j = 0; j < t->size(t->self); j++)
+        {
+            printf("%u\t", t->adjMat[i][j]);
+        }
+        printf("\n\n\n");
+    }
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_matrix.c:183]: 输出格式与其它语言不一致

File: codes/c/chapter_graph/graph_adjacency_matrix.c [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+void graphic(graphAdjMat* t, unsigned int number_vertices, int* vertices, unsigned int** adjMat)
+{    
+    //函数指针
+    t->addVertex = addVertex;
+    t->size = size;
+    t->addEdge = addEdge;
+    t->removeVertex = removeVertex;
+    t->removeEdge = removeEdge;
+    t->print_graph = print_graph;    
+    t->self = t;
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_matrix.c:196]: 不推荐此种写法，对新人来说会造成一定困惑

File: codes/c/chapter_graph/graph_adjacency_matrix.c [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+        for(int j = 0; j < t->size(t->self); j++)
+        {
+            printf("%u\t", t->adjMat[i][j]);
+        }
+        printf("\n\n\n");
+    }
+}
+
+/* 构造函数 */
+void graphic(graphAdjMat* t, unsigned int number_vertices, int* vertices, unsigned int** adjMat)
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_matrix.c:187]: 建议 `graphAdjMat* newGraphic(graphAdjMat* t,...)`

File: codes/c/chapter_graph/graph_adjacency_list.c [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+typedef struct node node;
+typedef struct graphAdjList graphAdjList;
+
+void free_Vertex(Vertex* t);
+
+/* 节点 */
+struct node
+{
+    Vertex* val;
+    node* next;
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_list.c:20]: 注释说明含义

File: codes/c/chapter_graph/graph_adjacency_list.c [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+};
+node* new_node()
+{
+    node* temp = (node*)malloc(sizeof(node));
+    temp->next = 0;
+    temp->val = 0;
+    return temp;
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_list.c:27]: `newNode`

File: codes/c/chapter_graph/graph_adjacency_list.c [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+ * File: graph_adjacency_list.c
+ * Created Time: 2023-07-07
+ * Author: NI-SW (947743645@qq.com)
+ */
+
+#include "../utils/common.h"
+
+typedef struct Vertex Vertex;
+typedef struct Link_list Link_list;    
+typedef struct node node;
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_list.c:11]: `typedef struct Node Node;`

File: codes/c/chapter_graph/graph_adjacency_list.c [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+typedef struct graphAdjList graphAdjList;
+
+void free_Vertex(Vertex* t);
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_list.c:14]: t 含义不明确

**Review by Gonglja — COMMENTED** [review]
整体存在以下几类问题哈

1. 与其它语言函数签名不一致
2. 缺少注释
3. 缩进风格（缩进可参考[14.1.3 C/C++ 环境](https://www.hello-algo.com/chapter_appendix/installation/#1413-cc)）、命名规则
4. 测试用例不一致

**Comment by NI-SW** [comment]
感谢指正，已修改XP

File: codes/c/chapter_graph/graph_adjacency_list.c [hunk 10]
```diff
@@ -0,0 +1,373 @@
+/**
+ * File: graph_adjacency_list.c
+ * Created Time: 2023-07-07
+ * Author: NI-SW (947743645@qq.com)
+ */
+
+#include "../utils/common.h"
+
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_list.c:8]: `typedef struct Vertex Vertex;`

File: codes/c/chapter_graph/graph_adjacency_list.c [hunk 11]
```diff
@@ -0,0 +1,373 @@
...
+void free_Vertex(struct Vertex* val);
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_list.c:9]: `void freeVertex(Vertex* val);`

File: codes/c/chapter_graph/graph_adjacency_matrix.c [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+    unsigned int** adjMat;
+    unsigned int size;
+    unsigned int capacity;
+};
+
+typedef struct graphAdjMat graphAdjMat;
+
+/* 添加边 */
+void addEdge(graphAdjMat* t, int i, int j)
+{
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_matrix.c:22]: 刷新一下格式，参见 https://www.hello-algo.com/chapter_appendix/installation/#1413-cc 步骤3
```c
void addEdge(graphAdjMat* t, int i, int j) {
    ...
}
```

File: codes/c/chapter_graph/graph_adjacency_matrix.c [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+    // 越界检查
+    if(i < 0 || j < 0 || i >= t->size || j >= t->size || i == j)
+    {
+        printf("Out of range in %s:%d\n", __FILE__, __LINE__);
+        exit(1);
+    }
+    
+    t->adjMat[i][j] = 1;
+    t->adjMat[j][i] = 1;
+
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_matrix.c:32]: 删除空行

File: codes/c/chapter_graph/graph_adjacency_matrix.c [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+    unsigned int capacity;
+};
+
+typedef struct graphAdjMat graphAdjMat;
+
+/* 添加边 */
+void addEdge(graphAdjMat* t, int i, int j)
+{
+    // 越界检查
+    if(i < 0 || j < 0 || i >= t->size || j >= t->size || i == j)
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_matrix.c:24]: `if (...)` `if` 后有空格

File: codes/c/chapter_graph/graph_adjacency_matrix.c [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+    }
+    
+    t->adjMat[i][j] = 1;
+    t->adjMat[j][i] = 1;
+
+}
+
+/* 删除边 */
+void removeEdge(graphAdjMat* t, int i, int j) 
+{
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_matrix.c:37]: 如上

File: codes/c/chapter_graph/graph_adjacency_matrix.c [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+    // 越界检查
+    if(i < 0 || j < 0 || i >= t->size || j >= t->size || i == j)
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_matrix.c:39]: 空格

File: codes/c/chapter_graph/graph_adjacency_matrix.c [hunk 17]
```diff
@@ -0,0 +1,280 @@
...
+    unsigned int** temp_Mat = (unsigned int**)malloc(sizeof(unsigned int*) * t->size * 2);
+    unsigned int* temp_Mat_line = (unsigned int*)malloc(sizeof(unsigned int) * (t->size * 2) * (t->size * 2));
+    memset(temp_Mat_line, 0,  sizeof(unsigned int) * (t->size * 2) * (t->size * 2));
+    for(int k = 0; k < t->size * 2; k++)
+    {
+        temp_Mat[k] = temp_Mat_line + k * (t->size * 2);
+    }
+    
+    // 原数据复制到新数组
+    for(int i = 0;i < t->size;i++)
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_matrix.c:86]: `for (int i = 0; i < t->size; i++)`

File: codes/c/chapter_graph/graph_adjacency_list.c [hunk 18]
```diff
@@ -0,0 +1,373 @@
...
+int main() 
+{
+
+    // 构造初始图结构
+    graphAdjList* graph = newGraphic();
+    addVertex(graph,4);    
+    addVertex(graph,5);
+    addVertex(graph,2);
+    addVertex(graph,3);
+    addVertex(graph,1);
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_list.c:338]: 顺序是否反了

File: codes/c/chapter_graph/graph_adjacency_list.c [hunk 19]
```diff
@@ -0,0 +1,373 @@
...
+{
+    // 链表节点内包含顶点类和下一个节点地址
+    struct Vertex* val;
+    struct Node* next;
+};
+
+typedef struct Node Node;
+
+/* 链表节点构造函数 */
+Node* new_node()
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_list.c:24]: `Node* newNode()`

File: codes/c/chapter_graph/graph_adjacency_list.c [hunk 20]
```diff
@@ -0,0 +1,373 @@
...
+    // 节点值
+    int val;
+    // 与其它节点相连接的边的链表
+    struct Link_list* linked;
+};
+
+typedef struct Vertex Vertex;
+
+/* 顶点节点构造函数 */
+Vertex* new_Vertex(int val)
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_list.c:44]: `Vertex* newVertex(int val)`

File: codes/c/chapter_graph/graph_adjacency_list.c [hunk 21]
```diff
@@ -0,0 +1,373 @@
...
+{
+    Vertex* v = (Vertex*)malloc(sizeof(Vertex));
+    // 为新节点赋值并建立该节点的链表
+    v->val = val;
+    v->linked = new_Linklist(v);
```
Comment (Gonglja) [inline codes/c/chapter_graph/graph_adjacency_list.c:49]: `newLinklist(v)`

**Review by Gonglja — COMMENTED** [review]
感谢 @NI-SW 

你还需要格式化下代码，调整一下命名

vscode 配置参考 https://www.hello-algo.com/chapter_appendix/installation/#1413-cc 中第三条

**Comment by NI-SW** [comment]
感谢@Gonglja
格式及命名都已修改 。

**Comment by Gonglja** [comment]
approve

**Review by krahets — APPROVED** [review]

**Pull request merged**